### DISCUSSION TITLE
How to use internal API to generate static page using getStaticProps

### DISCUSSION THREAD

**Discussion by mohammedsrehan** [vercel/next.js#38191] [Help] [answered]
How to use internal API to generate static page using getStaticProps. The docs specifies to not fetch the API but write the server logic directly into the getStaticProps. The problem is I'm not fetching data from an external API or a data file in the project, but I want to return the json object through the API route as defined in the next default example 

`pages/api/user.js`: 
```
export default function handler(req, res) {
  res.status(200).json({ name: 'John Doe' })
}
```
How can I use the data from this endpoint to generate a static page using getStaticProps. My use case has a larger object returned but for simplicity I have posted the example from the docs.

**Comment by icyJoseph** [comment, marked as answer]
You could use `process.cwd()` and the file read API to read the JSON directly inside `getStaticProps`, see here for more info: https://nextjs.org/docs/api-reference/data-fetching/get-static-props#reading-files-use-processcwd

The issue is not that Next.js wants to be annoying or anything like that, rather, at build time there's no server, and if there's no server running, then API routes are not active. Since these are all compiled at the same time, it is not reliable, nor trivial to wait for API routes to be compiled, setup a server and listen in  possible  requests  made from `getStaticProps`.

If you have a JSON file that has the data, in the same way that one would fetch from a 3rd party service, you can access that data from `getStaticProps`, by reading the file.

**Reply by mohammedsrehan** [reply]
> If you have a JSON file that has the data, in the same way that one would fetch from a 3rd party service, you can access that data from `getStaticProps`, by reading the file.

Can I fetch the data from the api file using `process.cwd` or I will have to create another file which has the data? ( In the latter use case, it will be of no use to have that api end point )

**Reply by icyJoseph** [reply]
No, you'd have to read it from a file. 

Another option:

```ts
// api/route
export const payload = { name: 'John Doe' };

export default function handler(req, res) {
  res.status(200).json(payload)
}
```

And then import the `payload` from `api/route`

```ts
import { payload } from './api/route';
```

> or I will have to create another file which has the data? ( In the latter use case, it will be of no use to have that api end point )

Yeah, then if there's no point for that API endpoint... 

I thought, perhaps you wanted to have an endpoint `/api/route` which you could reach from the browser, or virtually any other application, but if the JSON only matters for the `getStaticProps`, then forget about the API endpoint.

**Reply by icyJoseph** [reply]
If you are worried about leaking code to the client, please refer to code-elimination, as seen in this tool, https://next-code-elimination.vercel.app/, things used only in `getStaticProps` are removed from the client bundle.

**Reply by icyJoseph** [reply]
Don't forget to mark as answered if it did the trick for you!
